engine/src/main/java/org/terasology/engine/persistence/typeHandling/extensionTypes/ChunkMeshTypeHandler.java

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.engine.persistence.typeHandling.extensionTypes;
+
+import org.lwjgl.BufferUtils;
+import org.terasology.engine.registry.CoreRegistry;
+import org.terasology.engine.rendering.primitives.ChunkMesh;
+import org.terasology.engine.rendering.primitives.ChunkTessellator;
+import org.terasology.persistence.typeHandling.PersistedData;
+import org.terasology.persistence.typeHandling.PersistedDataSerializer;
+import org.terasology.persistence.typeHandling.TypeHandler;
+
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class ChunkMeshTypeHandler extends TypeHandler<ChunkMesh> {
+    @Override
+    protected PersistedData serializeNonNull(ChunkMesh value, PersistedDataSerializer serializer) {
+        if (!value.hasVertexElements()) {
+            throw new IllegalStateException("Attempting to serialize a ChunkMesh whose data has already been discarded.");
+        }
+        List<PersistedData> data = new ArrayList<>();
+        for (ChunkMesh.RenderType renderType : ChunkMesh.RenderType.values()) {
+            data.add(serializer.serialize(asByteBuffer(value.getVertexElements(renderType).finalVertices)));
+            data.add(serializer.serialize(asByteBuffer(value.getVertexElements(renderType).finalIndices)));
+        }
```

Comment:
this is quite a bit to serialize. any benefits with just encoding the block id?

Reply:
The mesh that a block generates depends on the context in which it occurs: all the adjacent blocks and all their lighting data usually, and sometimes the extraData fields too. In theory it could depend on the block entities too, although I don't know of any examples where that currently happens. In order to reconstruct the mesh, it would be necessary to put all that data into a `ChunkView` in order to re-generate the mesh, which would be pretty complicated.